gtdynamics/dynamics/Chain.cpp

```diff
@@ -0,0 +1,116 @@
+/* ----------------------------------------------------------------------------
+ * GTDynamics Copyright 2021, Georgia Tech Research Corporation,
+ * Atlanta, Georgia 30332-0415
+ * All Rights Reserved
+ * See LICENSE for the license information
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file Chain.cpp
+ * @brief Create a serial kinematic chain.
+ * @author Dan Barladeanu, Frank Dellaert.
+ */
+
+#include <gtdynamics/dynamics/Chain.h>
+
+namespace gtdynamics {
+
+void Chain::monoidOperation(const Pose3 &aTb, const Matrix &bAj,
+                            const Pose3 &bTc, const Matrix &cAk, 
+                            Pose3 &aTc, Matrix &cAjk) {
+    
+    // expect 6 rows in jacobians
+    if (bAj.rows() != 6 || cAk.rows() != 6) {
+      throw std::runtime_error(
+          "Jacobians should have 6 rows in SE(3)");
+    }
+
+    // Compose poses:
+    aTc = aTb.compose(bTc);
+    
+    // Adjoint the first Jacobian to the new end-effector frame C:
+    Matrix c_Ad_b = bTc.inverse().AdjointMap();
+
+    // Multiply with first chain screw axes:
+    Matrix mult = c_Ad_b * bAj;
+
+    // Create Matrix with correct number of columns:
+    Matrix out(mult.rows(), mult.cols() + cAk.cols());
+    
+    // Fill Matrix: 
+    out.leftCols(mult.cols()) = mult;
+    out.rightCols(cAk.cols()) = cAk;
+    
+    cAjk = out;
+}
+
+Chain Chain::compose(std::vector<Chain> &chain_vector){
+    Pose3 T, Ti, Tres;
+    Matrix A, Ai, Ares;
+
+    // Initialize from first chain in vector
+    T = chain_vector[0].sMb();
```

Comment:
Could be considerably simplified if we did overwrite operator* :-)

Reply:
done